pkg/controller/attach_detach.go

```diff
@@ -741,53 +734,64 @@ func (r *ReconcileAttachDetach) recoverAzVolumeAttachment(ctx context.Context, r
 		wg.Add(1)
 		go func(azv azdiskv1beta2.AzVolumeAttachment, azvMap *sync.Map) {
 			defer wg.Done()
+			var publishCtx map[string]string
+			if lun, _, err := r.cloudDiskAttacher.GetDiskLun(azv.Spec.VolumeName, azv.Spec.VolumeID, types.NodeName(azv.Spec.NodeName)); err == nil {
+				publishCtx = map[string]string{azureconstants.LUN: strconv.Itoa(int(lun))}
+			}
 			var targetState azdiskv1beta2.AzVolumeAttachmentAttachmentState
 			updateMode := azureutils.UpdateCRIStatus
-			if azv.Spec.RequestedRole == azdiskv1beta2.ReplicaRole {
-				updateMode = azureutils.UpdateAll
-			}
 			updateFunc := func(obj client.Object) error {
 				var err error
 				azv := obj.(*azdiskv1beta2.AzVolumeAttachment)
-				if azv.Spec.RequestedRole == azdiskv1beta2.ReplicaRole {
-					// conversion logic from v1beta1 to v1beta2 for replicas come here
-					azv.Status.Annotations = azv.ObjectMeta.Annotations
-					azv.ObjectMeta.Annotations = map[string]string{consts.VolumeAttachRequestAnnotation: "CRI recovery"}
-				}
+				azv = updateStatusDetail(azv, publishCtx)
 				// add a recover annotation to the CRI so that reconciliation can be triggered for the CRI even if CRI's current state == target state
 				azv.Status.Annotations = azureutils.AddToMap(azv.Status.Annotations, consts.RecoverAnnotation, recoveryID)
 				if azv.Status.State != targetState {
 					_, err = updateState(azv, targetState, forceUpdate)
 				}
 				return err
 			}
-			switch azv.Status.State {
-			case azdiskv1beta2.Attaching:
-				// reset state to Pending so Attach operation can be redone
-				targetState = azdiskv1beta2.AttachmentPending
-				innerUpdateFunc := updateFunc
-				updateFunc = func(obj client.Object) error {
-					if err := innerUpdateFunc(obj); err != nil {
-						return err
-					}
-					azv := obj.(*azdiskv1beta2.AzVolumeAttachment)
-					azv.Status.Detail = nil
-					azv.Status.Error = nil
-					return nil
-				}
-			case azdiskv1beta2.Detaching:
-				// reset state to Attached so Detach operation can be redone
+			shouldDelete := false
+			if attached, err := r.isAttached(ctx, azv.Spec.VolumeID, azv.Spec.NodeName); attached && err == nil {
 				targetState = azdiskv1beta2.Attached
-			default:
-				targetState = azv.Status.State
-			}
-
-			if _, err := azureutils.UpdateCRIWithRetry(ctx, nil, r.cachedClient, r.azClient, &azv, updateFunc, consts.ForcedUpdateMaxNetRetry, updateMode); err != nil {
-				w.Logger().Errorf(err, "failed to update AzVolumeAttachment (%s) for recovery", azv.Name)
-			} else {
-				// if update succeeded, add the CRI to the recoveryComplete list
+			} else if !attached && err == nil {
+				// disk is not attached. Delete the azVolumeAttachment
+				shouldDelete = true
+				w.Logger().Infof("need to detach AzVolumeAttachment (%s)", azv.Name)
+				markDetachRequest(&azv, attachdetach)
```

Comment:
I'm wondering should I use `updateFunc` in `UpdateCRIWithRetry()` to `markDetachRequest` instead of updating the object directly here.

Reply:
Modifying the object here likely will not result in a expected object update. As the update call actually happens in 788, where `UpdateCRIWithRetry` is called with the defined update func. Consider using below function to append to existing update func:
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/38590912a5baddba38b40cb94922346ff73592f3/pkg/azureutils/azure_disk_utils.go#L1254